ivadomed/utils.py

```diff
@@ -620,6 +640,115 @@ def threshold_predictions(predictions, thr=0.5):
+
+    # Compose transforms
+    do_transforms = compose_transforms(context['transformation_validation'])
+
+    # Undo Transforms
+    undo_transforms = ivadomed_transforms.UndoCompose(do_transforms)
+
+    # Load data
+    filename_pairs = [([fname_image], fname_image, fname_roi, [{}])]
+    if context['unet_3D'] == False:  # TODO: rename this param 'model_name' or 'kernel_dim'
+        ds = MRI2DSegmentationDataset(filename_pairs,
+                                      slice_axis=AXIS_DCT[context['slice_axis']],
+                                      cache=True,
+                                      transform=do_transforms,
+                                      slice_filter_fn=SliceFilter(**context["slice_filter"]),
+                                      canonical=True)
+    else:
+        print('\n3D unet is not implemented yet.')
+        exit()
+
+    # If fname_roi provided, then remove slices without ROI
+    if fname_roi is not None:
+        ds = ivadomed_loader.filter_roi(ds, nb_nonzero_thr=context["slice_filter_roi"])
+
+    if context['unet_3D'] == False:
+        print(f"\nLoaded {len(ds)} {context['slice_axis']} slices..")
+
+    # Data Loader
+    data_loader = DataLoader(ds, batch_size=context["batch_size"],
+                             shuffle=False, pin_memory=True,
+                             collate_fn=mt_datasets.mt_collate,
+                             num_workers=0)
+
+    # Load model
+    model = torch.load(fname_model, map_location=device)
+
+    # Inference time
+    model.eval()
+
+    # Loop across batches
+    preds_list, sliceIdx_list = [], []
+    for i_batch, batch in enumerate(data_loader):
+        with torch.no_grad():
+            preds = model(batch['input'])
+
+        rdict = {}
+        rdict['gt'] = preds
+        batch.update(rdict)
+
+        # Reconstruct 3D object
+        for i_slice in range(len(batch['gt'])):
+            # Undo transformations
+            rdict = {}
+            # Import transformations parameters
+            for k in batch.keys():
+                rdict[k] = batch[k][i_slice]
+            rdict_undo = undo_transforms(rdict)
+
+            # Add new segmented slice to preds_list
+            # Convert PIL to numpy
+            pred_cur = np.array(rdict_undo['gt'])
+            preds_list.append(pred_cur)
+            # Store the slice index of pred_cur in the original 3D image
+            sliceIdx_list.append(int(rdict_undo['input_metadata']['slice_index']))
+
+            # If last batch and last sample of this batch, then reconstruct 3D object
+            if i_batch == len(data_loader) - 1 and i_slice == len(batch['gt']) - 1:
+                pred_nib = pred_to_nib(data_lst=preds_list,
+                                       z_lst=sliceIdx_list,
+                                       fname_ref=fname_image,
+                                       fname_out=None,
```

Comment:
I am not sure to understand why the file is not saved here  (or given as an option in this function) since pred_to_nib allows to save files. The file will be returned than saved by an other function?

Reply:
Yes I think the idea was to choose the output file name in the SCT script

Reply:
`fname_out=None` --> not saved by `pred_to_nib` --> allow the user to play with the data eg:
- threshold it as it is soft seg at this stage
- do some home-made post-processing
- save the file how and where she/he wants